source/extensions/clusters/redis/redis_cluster.cc

```diff
@@ -0,0 +1,282 @@
+    ASSERT(!current_request_);
+    parent_.dispatcher_.deferredDelete(std::move(client_));
+  }
+}
+
+void RedisCluster::RedisDiscoverySession::registerDiscoveryAddress(
+    const std::list<Envoy::Network::Address::InstanceConstSharedPtr>& address_list) {
+  absl::MutexLock l(&discovery_mutex_);
+  discovery_address_list_.insert(discovery_address_list_.end(), address_list.begin(),
+                                 address_list.end());
+}
+
+void RedisCluster::RedisDiscoverySession::startResolve() {
+  absl::MutexLock l(&discovery_mutex_);
+  parent_.info_->stats().update_attempt_.inc();
+  // if a resolution is current in progress, skip
+  if (current_request_) {
+    return;
+  }
+
+  // if hosts is empty, we haven't got a successful result from CLUSTER SLOTS call yet, pick a
+  // random discovery address from dns
+  Upstream::HostSharedPtr host;
+  if (hosts_.empty()) {
+    const int rand_idx = parent_.random_.random() % discovery_address_list_.size();
+    auto it = discovery_address_list_.begin();
+    std::next(it, rand_idx);
+    host = Upstream::HostSharedPtr{new RedisHost(parent_.info(), "", *it, parent_, true)};
+  } else {
+    const int rand_idx = parent_.random_.random() % hosts_.size();
+    host = hosts_[rand_idx];
+  }
+
+  if (!client_) {
+    client_ = client_factory_.create(host, parent_.dispatcher_, *this);
+    client_->addConnectionCallbacks(*this);
+  }
+
+  current_request_ = client_->makeRequest(ClusterSlotsRequest::instance_, *this);
+}
+
+void RedisCluster::RedisDiscoverySession::onResponse(
+    NetworkFilters::Common::Redis::RespValuePtr&& value) {
+  absl::MutexLock l(&discovery_mutex_);
+
+  current_request_ = nullptr;
+  // TODO(hyang): move to use connection pool for the cluster and avoid closing the connection each
+  // time
+  if (client_) {
+    client_->close();
+  }
+
+  if (value->type() != NetworkFilters::Common::Redis::RespType::Array) {
+    ENVOY_LOG(error, "Unexpected response to cluster slot command: {}", value->toString());
+    parent_.info_->stats().update_no_rebuild_.inc();
+    return;
+  }
+
+  Upstream::HostVector new_hosts;
+  std::array<std::string, 16384> slots_;
+
+  for (const NetworkFilters::Common::Redis::RespValue& part : value->asArray()) {
+    const std::vector<NetworkFilters::Common::Redis::RespValue>& slotRange = part.asArray();
+    ASSERT(slotRange.size() >= 3);
+
+    // field 0: start slot range
+    auto startField = slotRange[0].asInteger();
+
+    // field 1: end slot range
+    auto endField = slotRange[1].asInteger();
+
+    // field 2: master address for slot range
+    // TODO(hyang): for now we're only adding the master node for each slot, when we're ready to
+    //  send requests to slave nodes, we need to add subsequent address in the response as slave
+    //  nodes
+    auto masterAddress = ProcessCluster(slotRange[2]);
+    // add to the slot
+    for (auto i = startField; i <= endField; ++i) {
+      slots_[i] = masterAddress->asString();
```

Comment:
The cluster_slots_map_ std::array() is a flexible approach for handling the slot mapping, but it does seem like it takes up a lot of memory (16384 std::string's) and CPU cycles (as seen here in the copying of 16384 strings over all cluster members) for a map that will generally scale as O(n) where n is the number of cluster nodes. I would suggest creating a // TODO around investigating the performance of alternative representations.

Reply:
Henry and I discussed this at length; we felt it was an acceptable mode of behavior for now but did not that this may not be ideal! I'll add a todo.

Reply:
Would it be possible to explain this mapping somewhere, maybe on the class comment? As someone that doesn't understand the protocol I think a brief description would be useful? I'm not sure what the tradeoffs here are.

Reply:
Will do